lib/OpenQA/Schema/Result/Needles.pm

```diff
@@ -103,30 +103,44 @@ sub update_needle_cache {
     }
 }
 
-# save the needle informations
+# save the needle information
 # be aware that giving the optional needle_cache hash ref, makes you responsible
 # to call update_needle_cache after a loop
 sub update_needle {
     my ($filename, $module, $matched, $needle_cache) = @_;
 
-    my $schema = OpenQA::Schema->singleton;
-    my $guard  = $schema->txn_scope_guard;
+    # assume that path of the JSON file is relative to the job's needle dir or (to support legacy versions
+    # of os-autoinst) relative to the "share dir" (the $bmwqemu::vars{PRJDIR} variable in legacy os-autoinst)
+    my $needle_dir;
+    unless (-f $filename) {
+        my $absolute_filename = catdir($needle_dir = $module->job->needle_dir, $filename);
+        my $needle_exists     = -f $absolute_filename;
 
-    if (!-f $filename) {
-        $filename = catdir($OpenQA::Utils::sharedir, $filename);
-        if (!-f $filename) {
-            log_error(
-                "Needle file $filename not found where expected. Check $OpenQA::Utils::prjdir for distri symlinks");
+        if (!$needle_exists) {
+            $absolute_filename = catdir($OpenQA::Utils::sharedir, $filename);
+            $needle_exists     = -f $absolute_filename;
+        }
+
+        if ($needle_exists) {
+            $filename = $absolute_filename;
+        }
+        else {
+            log_error("Needle file $filename not found within $needle_dir.");
+            return undef;
```

Comment:
would you mind save a bit on the whitespace-verbosity here e.g. by replacing L124-L130 with

```perl
return log_error("Needle file $filename not found within $needle_dir.") unless $needle_exists;
$filename = $absolute_filename;
```